Upgrade the Doctrine Coding Standard

nicwortel opened the pull request:
See https://github.com/doctrine/event-manager/pull/33, https://github.com/orgs/doctrine/projects/7#card-50248429.

In order to support PHP 8, the Coding Standard repository needs to be updated.

greg0ire commented on lib/Doctrine/Common/DataFixtures/Purger/ORMPurger.php:
```diff
@@ -190,7 +192,7 @@ private function getCommitOrder(EntityManagerInterface $em, array $classes)
 
                 /** @var ClassMetadata $targetClass */
-                $targetClass     = $em->getClassMetadata($assoc['targetEntity']);
+                $targetClass = $em->getClassMetadata($assoc['targetEntity']);
```
Both of these should have been fixed by `phpcbf` I think :thinking:

greg0ire commented:
You can add `<config name="php_version" value="70200"/>` towards the top, it might be helpful in avoiding some unwanted fixes.

nicwortel commented on phpcs.xml.dist:
```diff
@@ -14,13 +14,24 @@
+        <exclude name="SlevomatCodingStandard.TypeHints.ReturnTypeHint.MissingNativeTypeHint"/>
+        <exclude name="SlevomatCodingStandard.TypeHints.ParameterTypeHint.MissingNativeTypeHint"/>
+
+        <exclude name="SlevomatCodingStandard.TypeHints.ParameterTypeHint.MissingAnyTypeHint"/>
```
There are 27 missing parameter type hints. I think it is out of scope for this PR to fix them all. I suggest creating an issue to track this.

greg0ire replied:
I think I will rather take the time to contribute them to your PR. Feel free to start a separate PR before I do that though. I'd rather not commit "todo" things in the codebase.

nicwortel commented on phpcs.xml.dist:
```diff
@@ -14,13 +14,24 @@
+        <exclude name="SlevomatCodingStandard.TypeHints.ParameterTypeHint.MissingNativeTypeHint"/>
+
+        <exclude name="SlevomatCodingStandard.TypeHints.ParameterTypeHint.MissingAnyTypeHint"/>
+        <exclude name="SlevomatCodingStandard.TypeHints.ReturnTypeHint.MissingAnyTypeHint"/>
```
There are 54 missing return type hints. I think it is out of scope for this PR to fix them all. I suggest creating an issue to track this.

nicwortel commented on phpcs.xml.dist:
```diff
@@ -14,13 +14,24 @@
+
+        <exclude name="SlevomatCodingStandard.TypeHints.ParameterTypeHint.MissingAnyTypeHint"/>
+        <exclude name="SlevomatCodingStandard.TypeHints.ReturnTypeHint.MissingAnyTypeHint"/>
+        <exclude name="SlevomatCodingStandard.PHP.RequireExplicitAssertion.RequiredExplicitAssertion"/>
```
This sniff wants to automatically replace doc blocks with `assert` calls. This felt a little bit risky because this may actually break things.

greg0ire replied:
Not really, `assert` is not executed at all on sensibly configured production environments.

nicwortel commented:
Currently blocked because ODM doesn't allow PHP 8 yet, see https://github.com/doctrine/mongodb-odm/pull/2248.

greg0ire commented:
> Currently blocked because doctrine/mongodb-odm doesn't allow PHP 8 yet, see doctrine/mongodb-odm#2248.

Then let's focus on upgrading phpcs and only that, shall we?

nicwortel commented:
> Then let's focus on upgrading phpcs and only that, shall we?

I have changed the PR title and removed the last commit. The build is running again, but should turn green in a minute.

greg0ire reviewed (changes requested):
I'm going to checkout your PR and add the missing type hints

nicwortel commented on lib/Doctrine/Common/DataFixtures/ReferenceRepository.php:
```diff
@@ -199,6 +199,8 @@ public function getReferenceNames($reference)
      *
      * @param string $name
+     *
+     * @return bool
```
Any reasons for not using the native `: bool` return type?

greg0ire replied:
Yes: the class is not `final`, so that would be a BC-break.

nicwortel replied:
Good reason :smile:

nicwortel commented on tests/Doctrine/Tests/Common/DataFixtures/TestEntity/Link.php:
```diff
@@ -32,17 +32,17 @@ public function __construct(Uuid $id)
     }
 
-    public function getUrl()
+    public function getUrl(): string
```
`url` could actually be `null` because it's not initialized in the constructor.

But this is just a test fixture so I'm not sure how important this is.

nicwortel commented on tests/Doctrine/Tests/Common/DataFixtures/TestTypes/UuidType.php:
```diff
@@ -12,25 +12,38 @@ class UuidType extends Type
+    /**
+     * @param ?string $value
+     *
+     * @return ?Uuid
```
It should be possible to add this as a native return type hint.
